### PR TITLE
Fix dropping remote page in local buf bug

### PR DESCRIPTION
**Pull request by ctring**
When the local buffer pool is full, a page needs to be evicted from the buffer pool. This page will be written back to disk, this happens in the if block on line 286. 

We don't want/need to write a page of the remote relations to disk but the write back code initializes an smgr object that would write the page to disk anyway since it assumes all pages in the local buf to belong to temporary relations.

This PR changes that by storing extra metadata of local buffers in an array of `RemoteBufferDesc`s called `LocalBufferRemoteDescriptors`. Each `RemoteBufferDesc` can be seen as an extension of the corresponding `BufferDesc` struct in `LocalBufferDescriptors` (We don't want to add more fields in the `BufferDesc` struct because the shared buffer pool also uses this struct).  The `RemoteBufferDesc` store metadata that let us know whether a buffer is storing a page of a remote relation (`is_remote`) and the local transaction id (`lxid`). The local transaction id field was in `LocalBufferLookupEnt` previously but we move it here now to make things more organized.

If the `is_remote` field of a buffer is `true`, we simply discard the page instead of writing it back to disk.

### CODE REVIEW

File: src/backend/storage/buffer/localbuf.c [hunk 1]
```diff
@@ -279,35 +292,45 @@ LocalBufferAlloc(SMgrRelation smgr, ForkNumber forkNum, BlockNumber blockNum,
...
-		/* Find smgr relation for buffer */
-		if (am_wal_redo_postgres && MyBackendId == InvalidBackendId)
-			oreln = smgropen(bufHdr->tag.rnode, MyBackendId, RELPERSISTENCE_PERMANENT, UNKNOWN_REGION);
-		else
-			oreln = smgropen(bufHdr->tag.rnode, MyBackendId, RELPERSISTENCE_TEMP, UNKNOWN_REGION);
-
-		PageSetChecksumInplace(localpage, bufHdr->tag.blockNum);
+		/*
+		 * Remotexact
+		 * Just throw away the page if it comes from a remote relation
```
Comment (poojanilangekar) [inline src/backend/storage/buffer/localbuf.c]: I would reword it as write localbuf to disk only if it does not belong to a remote relation.

File: src/backend/storage/buffer/localbuf.c [hunk 2]
```diff
@@ -514,7 +539,10 @@ InitLocalBuffers(void)
 	LocalBufferDescriptors = (BufferDesc *) calloc(nbufs, sizeof(BufferDesc));
 	LocalBufferBlockPointers = (Block *) calloc(nbufs, sizeof(Block));
 	LocalRefCount = (int32 *) calloc(nbufs, sizeof(int32));
-	if (!LocalBufferDescriptors || !LocalBufferBlockPointers || !LocalRefCount)
+	/* Remotexact */
+	LocalBufferRemoteDescriptors = (RemoteBufferDesc *) calloc(nbufs, sizeof(RemoteBufferDesc));
```
Comment (poojanilangekar) [inline src/backend/storage/buffer/localbuf.c:555]: Should we increase nbufs now that we won't be writing some local bufs to disk?

**Review by poojanilangekar — APPROVED** [review]
Thanks you!

File: src/backend/storage/buffer/localbuf.c (hunk 1 above)
Comment (ctring) [reply src/backend/storage/buffer/localbuf.c]: Done

File: src/backend/storage/buffer/localbuf.c (hunk 2 above)
Comment (ctring) [reply src/backend/storage/buffer/localbuf.c:555]: One nice thing about local buffer is that you can increase the number of buffers without much overhead, and only transactions that need the extra buffers would allocate memory for the pages. 

> A session will allocate temporary buffers as needed up to the limit given by temp_buffers. The cost of setting a large value in sessions that do not actually need many temporary buffers is only a buffer descriptor, or about 64 bytes, per increment in temp_buffers. However if a buffer is actually used an additional 8192 bytes will be consumed for it (or in general, BLCKSZ bytes).

from `temp_buffers` in https://www.postgresql.org/docs/current/runtime-config-resource.html 

So, I think we can just crank the number of buffers up for now while reducing the size of the batch when loading data. I'll also put a warning (or error?) to make it clear that we are throwing a modified page if that ever happens.

File: src/backend/storage/buffer/localbuf.c (hunk 1 above)
Comment (ctring) [reply src/backend/storage/buffer/localbuf.c:306]: Actually, I can just skip the remote pages in the clock sweep run, and the `no empty local buffer available` error will be triggered when we don't have enough buffer

File: src/backend/storage/buffer/localbuf.c (hunk 2 above)
Comment (ctring) [reply src/backend/storage/buffer/localbuf.c:555]: Actually, I can just skip the remote pages in the clock sweep run, and the `no empty local buffer available` error will be triggered when we don't have enough buffers.

Comment (ctring) [reply src/backend/storage/buffer/localbuf.c:555]: Updated

**Comment by ctring** [comment]
Added additional checking regarding who owns the buffer.

**Comment by ctring** [comment]
Will add the hint bit optimization in the next PR.

File: src/backend/storage/buffer/localbuf.c [hunk 3]
```diff
@@ -258,6 +272,18 @@ LocalBufferAlloc(SMgrRelation smgr, ForkNumber forkNum, BlockNumber blockNum,
...
+				remote_bufHdr->is_remote &&
+				remote_bufHdr->lxid == MyProc->lxid)
+			{
+				/*
+				 * Remotexact
+				 * Prevent eviction of buffers that are owned by the current transaction
+				 * and contain dirty remote pages.
+				 */
+				continue;
+			}
```
Comment (ctring) [inline src/backend/storage/buffer/localbuf.c:285]: This is the latest change

**Review by poojanilangekar — APPROVED** [review]
